### PR TITLE
Bug 1876440: Cloud init section of common-templates is ignored

### PR DESCRIPTION
**Pull request by glekner**
image

### CODE REVIEW

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/vm-settings-tab-state-update.ts [hunk 1]
```diff
@@ -115,6 +126,40 @@ const osUpdater = ({ id, prevState, dispatch, getState }: UpdateOptions) => {
       { isHidden: asHidden(!isWindows, VMSettingsField.OPERATING_SYSTEM), value: isWindows },
     ),
   );
+
+  const relevantOptions = iGetRelevantTemplateSelectors(state, id);
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/vm-settings-tab-state-update.ts:130]: relevant template can change on relevenatOptions, not just os. IMO this needs its own updater and possibly new file called eg commonTemplatesPrefil

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/vm-settings-tab-state-update.ts [hunk 2]
```diff
@@ -115,6 +126,40 @@ const osUpdater = ({ id, prevState, dispatch, getState }: UpdateOptions) => {
...
+  const iCommonTemplates = iGetLoadedCommonData(state, id, VMWizardProps.commonTemplates);
+  const iTemplate =
+    iCommonTemplates && iGetRelevantTemplate(null, iCommonTemplates, relevantOptions);
+  const iCloudInitVolume = iGetCommonTemplateCloudInit(iTemplate);
+  const [data, isBase64] = CloudInitDataHelper.getUserData(
+    toShallowJS(iCloudInitVolume)?.cloudInitNoCloud,
+  );
+
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/vm-settings-tab-state-update.ts:138]: please take an inspiration from prefill-vm-template-state-update.ts

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/vm-settings-tab-state-update.ts [hunk 3]
```diff
@@ -115,6 +126,40 @@ const osUpdater = ({ id, prevState, dispatch, getState }: UpdateOptions) => {
...
+  if (data) {
+    dispatch(
+      vmWizardInternalActions[InternalActionType.SetCloudInitFieldValue](
+        id,
+        CloudInitField.IS_FORM,
+        false,
+      ),
+    );
+    dispatch(
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/vm-settings-tab-state-update.ts:147]: you need to take into account cloudinit from other templates.

Comment (suomiy) [reply frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/vm-settings-tab-state-update.ts:147]: also please test this with a userTemplate

File: frontend/packages/kubevirt-plugin/src/selectors/immutable/template/combined.ts [hunk 4]
```diff
@@ -97,3 +98,8 @@ export const iGetRelevantTemplates = (
 export const iGetRelevantTemplate = (
   ...args: Parameters<typeof iGetRelevantTemplates>
 ): ITemplate => iGetRelevantTemplates(...args).first();
+
+export const iGetCommonTemplateCloudInit = (tmp: ITemplate) =>
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/selectors/immutable/template/combined.ts:102]: please make an immutable version of `selectVM`. Specifying an index is unreliable.

**Review by suomiy — CHANGES_REQUESTED** [review]

**Comment by yaacov** [comment]
Note:
When using this I get double "cloundinit" disks when I change the "OS" input, and then change it again ...

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/vm-settings-tab-state-update.ts (hunk 1 above)
Comment (suomiy) [reply frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/vm-settings-tab-state-update.ts:130]: as mentioned here https://bugzilla.redhat.com/show_bug.cgi?id=1874074.

We are going to need a generic approach to common template prefill soon.

File: frontend/packages/kubevirt-plugin/src/selectors/immutable/template/combined.ts (hunk 4 above)
Comment (glekner) [reply frontend/packages/kubevirt-plugin/src/selectors/immutable/template/combined.ts:96]: created

**Comment by glekner** [comment]
* Handling of cloudinit disk and volume now works fine when dynamically changing os,flavor,workload when using common templates.
* I noticed when using a user template, cloud init isn't imported? is this bug known? @jelkosz @suomiy @yaacov

**Comment by yaacov** [comment]
> I noticed when using a user template, cloud init isn't imported? is this bug known? @jelkosz @suomiy @yaacov

you need to listen to changes in the user template, currently you only listen to common template fields.
but the user template issue can be tracked using a different bug report ( unless you want to fix it using this one ? ).

**Pull request merged**